src/merge_events.cpp

```diff
@@ -0,0 +1,205 @@
+#include <set>
+#include <vector>
+
+#include <backends.hpp>
+#include <cell_group.hpp>
+#include <cell_group_factory.hpp>
+#include <domain_decomposition.hpp>
+#include <merge_events.hpp>
+#include <model.hpp>
+#include <recipe.hpp>
+#include <util/filter.hpp>
+#include <util/span.hpp>
+#include <util/unique_any.hpp>
+#include <profiling/profiler.hpp>
+
+namespace arb {
+
+namespace impl {
+
+// The tournament tree data structure is used to merge k sorted lists of events.
+// See online for high-level information about tournament trees.
+//
+// This implementation maintains a heap-like data structure, with entries of type:
+//      std::pair<unsigned, post_synaptic_event>
+// where the unsigned ∈ [0, k-1] is the id of the list from which the event was
+// drawn. The id is stored so that the operation of removing the most recent event
+// knows which leaf node needs to be updated (i.e. the leaf node of the list from
+// which the most recent event was drawn).
+//
+// unsigned is used for storing the index, because if drawing events from more
+// event generators than can be counted using an unsigned a complete redesign
+// will be needed.
+tourney_tree::tourney_tree(std::vector<event_generator_ptr>& input):
+    input_(input),
+    n_lanes_(input_.size())
+{
+    // Must have at least 1 queue
+    EXPECTS(n_lanes_);
+    // Maximum value in unsigned limits how many queues we can have
+    EXPECTS(n_lanes_<(1u<<(sizeof(unsigned)*8u-1u)));
+
+    leaves_ = next_power_2(n_lanes_);
+    nodes_ = 2u*(leaves_-1u)+1u; // 2*l-1 with overflow protection
+
+    // Allocate space for the tree nodes
+    heap_.resize(nodes_);
+    // Set the leaf nodes
+    for (auto i=0u; i<leaves_; ++i) {
+        heap_[leaf(i)] = i<n_lanes_?
+            key_val(i, input[i]->next()):
+            key_val(i, terminal_pse()); // null leaf node
+    }
+    // Walk the tree to initialize the non-leaf nodes
+    setup(0);
+}
+
+void tourney_tree::print() const {
+    auto nxt=1u;
+    for (auto i=0u; i<nodes_; ++i) {
+        if (i==nxt-1) { nxt*=2; std::cout << "\n";}
+        std::cout << "{" << heap_[i].first << "," << heap_[i].second << "}\n";
+    }
+}
+
+bool tourney_tree::empty() const {
+    return event(0).time == max_time;
+}
+
+bool tourney_tree::empty(time_type t) const {
+    return event(0).time >= t;
+}
+
+postsynaptic_spike_event tourney_tree::head() const {
+    return event(0);
+}
+
+// Remove the smallest (most recent) event from the tree, then update the
```

Comment:
This explanation talks about the tree, but the implementation pops from the input lane

Reply:
It removes an event from the appropriate input lane, places it at a leaf node of the tree, then re-heapifys the tree.
This is all predicated on someone understanding what a tournament tree. I am not inclined to fully document the algorithm, because this is used in just one spot, and not exported for use elsewhere (well, it is exported in the `impl` namespace so that it can be unit tested).